metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -354,19 +380,106 @@ final class FileDecoderProvider(
 
   private def decodeFromTastyFile(
       pathInfo: PathInfo
-  ): Future[Option[String]] = {
+  ): Future[Option[String]] =
+    loadPresentationCompiler(pathInfo) match {
+      case Some(pc) =>
+        pc.getTasty(pathInfo.path.toURI, false, false).asScala.map(Some(_))
+      case None =>
+        Future.successful(None)
+    }
+
+  private def loadPresentationCompiler(
+      pathInfo: PathInfo
+  ): Option[PresentationCompiler] =
     for {
-      pc <- Future {
-        for {
-          targetId <- pathInfo.targetId
-          pc <- compilers.loadCompiler(targetId)
-        } yield pc
+      targetId <- pathInfo.targetId
+      pc <- compilers.loadCompiler(targetId)
+    } yield pc
+
+  /**
+   * For clients supporting executing commands [[TastyResponse]] is returned and clients can determine on their own how to handle returned value.
+   * If client supports http, he is redirected to the tasty endpoint defined at [[MetalsHttpServer]].
+   * That endpoint reuses logic declared in [[TastyHandler]]
+   * In both cases logic is pretty same:
+   * - for a given URI (which could be .scala or .tasty file itself) try to find .tasty file
+   * - dispatch request to the Presentation Compiler. It's worth noting that PC takes into account
+   *   client configuration to determine proper response format (HTML, console or plain text)
+   */
+  def executeShowTastyCommand(uri: URI, position: l.Position): Future[Unit] =
+    if (
+      clientConfig.isExecuteClientCommandProvider() && !clientConfig
+        .isHttpEnabled()
+    ) {
+      val response = getTastyForURI(uri).map { result =>
+        DecoderResponse(
+          uri.toString(),
+          result.fold(_ => null, identity),
+          result.fold(identity, _ => null)
+        )
       }
-      output <- pc match {
-        case None => Future.successful(None)
-        case Some(pc) =>
-          pc.getTasty(pathInfo.path.toURI, false, false).asScala.map(Some(_))
+      response.map { tasty =>
+        val command = new l.ExecuteCommandParams(
+          "metals-show-tasty",
+          List[Object](tasty).asJava
+        )
+        languageClient.metalsExecuteClientCommand(command)
+        scribe.debug(s"Executing show TASTy $command")
+      }
+    } else
+      httpServer() match {
+        case Some(server) =>
+          Future.successful(
+            Urls.openBrowser(server.address + s"/tasty?file=$uri")
+          )
+        case None =>
+          Future.successful {
+            scribe.warn(
+              "Unable to run show tasty. Make sure `isHttpEnabled` is set to `true`."
+            )
+          }
       }
-    } yield output
+
+  def getTastyForURI(
+      uri: URI,
+      cursorPosition: Option[l.Position] = None
+  ): Future[Either[String, String]] = {
+    val position = cursorPosition.getOrElse(new l.Position(1, 1))
```

Comment:
This will most likely not work if packages are in weird places, we should instead just check the package inside the tree and create a class file name based on that.